src/basic-noise-qvm.lisp

```diff
@@ -0,0 +1,194 @@
+   :t1-ops (make-hash-table :test 'eql)
+   :t2-ops (make-hash-table :test 'eql)
+   :depol-ops (make-hash-table :test 'eql)
+   :avg-gate-time (error ":AVG-GATE-TIME is a required initarg to BASIC-NOISE-QVM")
+   :readout-povms (make-hash-table :test 'eql))
+  (:documentation "BASIC-NOISE-QVM is a QVM that supports a noise model defined by T1, T2, and readout fidelities on each qubit. At each instruction, T1 noise and T2 noise is applied for the qubits involved in the instruction. Upon MEASURE, the readout noise is also applied for whichever qubit is being measured."))
+
+
+(defmethod initialize-instance :after ((qvm basic-noise-qvm) &rest args)
+  ;; If an basic-noise-qvm is initialized with T1, T2, or READOUT-POVM
+  ;; values, check that they are valid.
+  (declare (ignore args))
+  (maphash #'%check-noise-entry (t1-ops qvm))
+  (maphash #'%check-noise-entry (t2-ops qvm))
+  (maphash #'%check-noise-entry (depol-ops qvm))
+  (maphash #'%check-povm-entry (readout-povms qvm)))
+
+
+(defun %check-noise-entry (qubit noise-ops)
+  "Helper function to check that a key value pair in T1-OPS or T2-OPS is valid. The qubit must be a NON-NEGATIVE INTEGER, and the NOISE-OPS must be a valid kraus map."
+  (assert (and (integerp qubit) (<= 0 qubit)))
+  (check-kraus-ops noise-ops))
+
+
+(defun %check-povm-entry (qubit povm)
+  "Helper fucntion to check that a key value pair in READOUT-POVMS is valid. The QUBIT must be a NON-NEGATIVE INTEGER, and the POVM must be a correctly formatted POVM."
+  (assert (and (integerp qubit) (<= 0 qubit)))
+  (check-povm povm))
+
+
+(defun (setf qubit-t1) (t1 qvm qubit)
+  "Save the kraus operators for the T1 provided on the specific qubit."
+  (let ((kraus-ops (generate-damping-kraus-map t1 (avg-gate-time qvm))))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (t1-ops qvm)) kraus-ops)))
+
+
+(defun (setf qubit-t2) (t2 qvm qubit)
+  "Save the kraus operators for the T2  provided on the specified qubit."
+  (let ((kraus-ops (generate-dephasing-kraus-map t2 (avg-gate-time qvm))))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (t2-ops qvm)) kraus-ops)))
+
+
+(defun (setf qubit-depolarization) (depol-prob qvm qubit)
+  "Save the kraus operators for the depolarizing channel on QUBIT. The depolarizing channel represents the average gate overrotation/underrotation noise. It is specified by DEPOL-PROB, which is the probability of overrotation/underrotation for the given QUBIT."
+  (let ((kraus-ops (generate-depolarizing-kraus-map depol-prob)))
+    (check-kraus-ops kraus-ops)
+    (setf (gethash qubit (depol-ops qvm)) kraus-ops)))
+
+
+(defun (setf qubit-fro) (fro qvm qubit)
+  "Save the readout povms for the fro provided on the specified qubit. The fR0 is assumed to be the average readout fidelity (avg of p(1|1) and p(0|0)) so as an approximation, we set p(0|0) = p(1|1) = fR0 in the readout povm assignment probabilities."
+  (setf (gethash qubit (readout-povms qvm)) (list fro (- 1.0d0 fro) (- 1.0d0 fro) fro)))
```

Comment:
This produces slightly better code: 
```suggestion
(let ((one-minus-fro (- 1.0d0 fro)))
  (setf (gethash qubit (readout-povms qvm)) (list fro one-minus-fro one-minus-fro fro)))
```
One would think that the compiler ought to notice that (- 1.0d0 fro) is constant and reuse the value but, alas, it computes it twice unless you introduce the `let` form.